src/compiler/transformers/es2015.ts

```diff
@@ -980,40 +980,40 @@ namespace ts {
             const statements: Statement[] = [];
             resumeLexicalEnvironment();
 
+            // In derived classes, there may be code before the necessary super() call

```

Comment:
Part of the existing complexity here is my fault. The goal was to produce a statement list that contained the following statements in a specific order:

1. Existing *standard* prologue directives (i.e., `"use strict";`) from the original statement list.
2. New *custom* prologue statements (not really prologue directives, really just code injected into the top of the function body)
3. Existing *custom* prologue statements (again, not true prologue directives, but code injection used by transformations.
4. (new) Statements that come before `super()`
5. Injected code for `this` capturing as it applies to `super()` invocation.
6. The `super()` invocation itself.
7. Any remaining statements.

Also, the `mergeLexicalEnvironment` call injects temporary variable declarations between (2) and (3) above.